proto/keyserverconfig.proto

```diff
@@ -119,24 +119,114 @@ message KeyserverConfig {
-	// accepted. This option can be used to allow proofs emailed directly to
-	// the keyserver to be accepted (but the keyserver does NOT implement a
-	// SMTP interface right now).
+	// # Registration policy
+	// Specifying a registration policy inline in this configuration is
+	// DEPRECATED. Use RegistrationPolicy (@12) instead.  In
+	// versions that support the RegistrationPolicy, any value (except
+	// for the default) in there takes precedence over every inline
+	// configuration. No new configuration should be deployed with inline
+	// configuration set. 
+	// Those currently relying on this option should:
+	// 1) update keyserver to a version that supports @12
+	// 2) change the configuration file to specify @12
+	// 3) remove the value of this option from the configuration file
+
+	// EmailProofToAddr (DEPRECATED) specifies the additional allowed to
+	// address in email proofs. By default, only proofs sent to the user being
+	// registered all accepted. This option can be used to allow proofs emailed
+	// directly to the keyserver to be accepted (but the keyserver does NOT
+	// implement a SMTP interface right now).
 	string email_proof_to_addr = 8;
-	// EmailProofSubjectPrefix is used for DKIM-based email address
-	// registration.  The proof challenge is sent in the subject line, with an
-	// optional string preceding it. For example, if EmailProofSubjectPrefix =
-	// "account verification: ", then the proof email needs to have a subject
-	// line "account verification: abcd" for verify challenge abcd.
+	// EmailProofSubjectPrefix (DEPRECATED) is used for DKIM-based email
+	// address registration.  The proof challenge is sent in the subject line,
+	// with an optional string preceding it. For example, if
+	// EmailProofSubjectPrefix = "account verification: ", then the proof email
+	// needs to have a subject line "account verification: abcd" for verify
+	// challenge abcd.
 	string email_proof_subject_prefix = 9;
-
-	// EmailProofAllowedDomains specifies the domains for which this keyserver
-	// accepts email address registrations.
+	// EmailProofAllowedDomains (DEPRECATED) specifies the domains for which
+	// this keyserver accepts email address registrations.
 	repeated string email_proof_allowed_domains = 10;
 
 	bool insecure_skip_email_proof = 11;
+
+	// Keyserver may support multiple registration policies at a time.
+	// A policy is acceptable only for the domains it supports.
+	repeated RegistrationPolicy registration_policy = 12;
+}
+
+// RegistrationPolicy specifies the list of policies the keyserver may support,
+// to confirm the ownership of an email address
+message RegistrationPolicy {
+	oneof policy_type {
+		EmailProofByDKIM email_proof_by_dkim = 1	[(gogoproto.customname) = "EmailProofByDKIM"];
+		EmailProofByClientCert email_proof_by_client_cert = 2;
+		EmailProofByOIDC email_proof_by_oidc = 3	[(gogoproto.customname) = "EmailProofByOIDC"];
+		EmailProofByCustomAuth email_proof_by_custom_auth = 4;
```

Comment:
Again, I don't see how declaring 4 as "custom" helps us. Not interpreting an unknown registration policy at all still lets the server to error out because there is not a registration policy it can interpret.